### PR TITLE
[RFC] wm4 commits

### PR DESCRIPTION
**Pull request by xantoz**
Just a plain rebase (+ cherry-picks to keep it up to date) of https://github.com/wm4/mpv_backup master onto mpv master. Some manual edits was needed to resolve conflicts.

### CODE REVIEW

**Comment by CounterPillow** [comment]
Some people may be a little mad about the removal of optical disc support.

**Comment by xantoz** [comment]
@CounterPillow yes, that might be one of the more controversial changes here. I can try to make a rebase omitting those commits if we feel like we want to keep optical disc and TV support (the latter is still usable via av:// however) no matter what.

Although when making this I felt that trying to exclude those commits might complicate incorporating any future changes.

**Comment by olifre** [comment]
I certainly also have strong feelings about dropping DVB support. I've basically maintained it and heavily use it myself, it has seen regular contributions, and a recent overhaul by @rozhuk-im . While I agree the `tv://` stuff was rather borked (and I don't believe there are many users out there), this was not really true for the `dvb://` code. It was working very well across most DVB cards out there. The commit dropping DVB support only talks about the ancient TV code, though, so I wonder if it really was dropped on purpose. 

No strong feelings about optical disc support though on my end.

**Comment by rozhuk-im** [comment]
I use analog input few times while debug some device with analog video output.
My country have many TV channels in analog, and I think many countries have analog TV.
If there is some problems/issues with analog TV - I can try to see it, I steel have one analog TV tuner.

**Comment by qyot27** [comment]
I guess the real question should be: how much does mpv's internal usage of libcdio/cdparanoia and libbluray differ from FFmpeg's (there is currently no support in FFmpeg for dvdread/dvdnav, [although there were some patches on FFmpeg-devel way back in 2009](http://ffmpeg.org/pipermail/ffmpeg-devel/2009-July/069403.html))?  Ideally, if mpv handles something better, and it's something that could be added to FFmpeg's wrapper(s) for those libraries, it should.  If it's mpv-specific, how much of it is necessary to bridge the difference?  How much is just straight-up duplicated?

`av://` can be used to play audio CDs through FFmpeg, but like FFmpeg, it requires mpv to seek in the stream in order to initiate playback:
```
mpv --start=00:00:00.000000 av://libcdio:h:
```
It does seem to have one advantage over mpv's `cdda://` protocol: FFmpeg can detect the total length of the disc and seek in it immediately, the way we'd expect to do.  mpv starts caching the CD piecemeal, and can only seek as far as it's cached.  Unless I just missed where this is possible, mpv doesn't seem to support skipping to the next track (neither does FFmpeg, but it does display the information on the info readout as chapters).

Not sure if anything about `lavf://` changed in this series of commits, but with upstream mpv, at least libbluray works through it (albeit without metadata and chapter information, unless there's a way libavformat's API exposes those).  On Windows the invocation is a bit funky-looking, but it works:
```
E:\>mpv -vo direct3d --end 00:00:01 lavf://bluray:i:
Playing: lavf://bluray:i:
../../src/libbluray/bdnav/index_parse.c:227: index.bdmv: unknown extension data at 276
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
../../src/libbluray/bdnav/mpls_parse.c:983: _parse_mpls_extension(): unhandled extension 3.5
[ffmpeg/demuxer] mpegts: start time for stream 5 is not set in estimate_timings_from_pts
[ffmpeg/demuxer] mpegts: start time for stream 6 is not set in estimate_timings_from_pts
[ffmpeg/demuxer] mpegts: start time for stream 7 is not set in estimate_timings_from_pts
[ffmpeg/demuxer] mpegts: start time for stream 8 is not set in estimate_timings_from_pts
../../src/libbluray/bluray.c:256: 00000.m2ts: no timestamp for SPN 0 (got 0). clip 188955000-476758766.
[ffmpeg/demuxer] mpegts: Could not find codec parameters for stream 5 (Subtitle: hdmv_pgs_subtitle (pgssub) ([144][0][0]
[0] / 0x0090)): unspecified size
[ffmpeg/demuxer] Consider increasing the value for the 'analyzeduration' and 'probesize' options
[ffmpeg/demuxer] mpegts: Could not find codec parameters for stream 6 (Subtitle: hdmv_pgs_subtitle (pgssub) ([144][0][0]
[0] / 0x0090)): unspecified size
[ffmpeg/demuxer] Consider increasing the value for the 'analyzeduration' and 'probesize' options
[ffmpeg/demuxer] mpegts: Could not find codec parameters for stream 7 (Subtitle: hdmv_pgs_subtitle (pgssub) ([144][0][0]
[0] / 0x0090)): unspecified size
[ffmpeg/demuxer] Consider increasing the value for the 'analyzeduration' and 'probesize' options
[ffmpeg/demuxer] mpegts: Could not find codec parameters for stream 8 (Subtitle: hdmv_pgs_subtitle (pgssub) ([144][0][0]
[0] / 0x0090)): unspecified size
[ffmpeg/demuxer] Consider increasing the value for the 'analyzeduration' and 'probesize' options
 (+) Video --vid=1 (hevc 3840x2160 23.976fps)
 (+) Audio --aid=1 (dts 6ch 48000Hz)
     Audio --aid=2 (dts 2ch 48000Hz)
     Audio --aid=3 (dts 6ch 48000Hz)
     Audio --aid=4 (dts 2ch 48000Hz)
     Subs  --sid=1 (hdmv_pgs_subtitle)
     Subs  --sid=2 (hdmv_pgs_subtitle)
     Subs  --sid=3 (hdmv_pgs_subtitle)
     Subs  --sid=4 (hdmv_pgs_subtitle)
AO: [wasapi] 48000Hz 5.1(side) 6ch s32
VO: [direct3d] 3840x2160 yuv420p10
[osd/libass] fontselect: Using default font: (sans-serif, 400, 0) -> E:/Programs/mpv-player/mpv/subfont.ttf, 0, (none)
AV: 00:00:00 / 01:46:34 (0%) A-V:  0.000 Cache: 19s+160MB


Exiting... (End of file)

E:\>
```

I'd just suggest - in the event that the optical disc support is removed - that we might want to offer a shortcut to enable playback through FFmpeg, like `--bluray` or `--audiocd` (either standalone or as a replacement for both the protocol invocation and `--*-device=<location>` option), rather than try to have users remember to invoke the convoluted FFmpeg passthrough syntax for whichever type of disc you're trying to access.


Personally, out of the three, I'd be most likely to miss the support for libbluray.  I never play CDs directly, and DVDs are small enough to warrant just dumping the entire rip to the hard drive (at which point I'm probably not going to be playing it, I'll be converting it to some other format first).  But Blu-ray can stress storage space far more than DVD ever could, so playing off the disc is still a viable option if you've A1) got the VUK/Unit keys with upstream libaacs or A2) are using MakeMKV's libmmbd->libaacs compatibility mode, or B) had at one time made a decrypted 1:1 backup.  mpv's default invocation is just a lot more sane than FFmpeg's, and easier to figure out.  So really, if it can get by simply redirecting through libavformat for optical discs, a sane treatment and easy-to-remember way to activate it are all I really care about.


tl;dr:
FFmpeg seems to support audio CDs better than mpv does, Blu-ray seems like it might be a draw.  mpv supports DVD, FFmpeg doesn't.   FFmpeg seems to export CD track timecodes as chapters, it cannot do the same for Blu-ray, because it's a protocol and not a demuxer.  mpv doesn't(?) get CD track timecodes using `cdda://`, but since FFmpeg uses libcdio as a demuxer, it might be possible.  mpv gets metadata and chapter information from libbluray.

Basically, it looks highly preferable to switch to FFmpeg for audio CDs (with a little grease so that it's not necessary to explicitly seek), and it opens up the possibility that next/previous track seeking could be available with better use of libavformat's API.  Can we use libbluray selectively, just to get the metadata and chapter info, but rely on FFmpeg to do everything else?  DVD looks like an impasse, unless the patches for libavformat get rehabilitated and merged in.

**Comment by kkkrackpot** [comment]
If killing optical stuff support will also kill the ability to play ripped DVD/BD/UHD structures (i.e. folders on HDD with STREAM, etc. in it) directly (i.e. like `mpv /path/to/structure`), without recoding to MKVs -- from user's point of view the benefit of this removal would be unclear...

**Comment by xantoz** [comment]
Turns out that many later refactoring commits depend more or less on the big removal commit, making simply not picking it rather difficult. There are many changes after it which would need to be applied to files which were removed in that commit, for instance, and others where we need to keep some functionality that has been possible to simplify away much in part due to the removal of things like DVD support.

If something needs to be kept it's probably best to try to single out a few things most worth saving, favoring still removing as much as possible and instead, as @qyot27 points out, trying to find ways to switch over to using ffmpeg functionality where feasible, possibly with some user-friendly short-cuts.

**Comment by dnmTX** [comment]
@wm4  have a question about #6127 : 
This is what i use at the moment: 
`cache-initial=0
cache-pause-initial=yes
cache-pause-wait=5.000
cache-file=TMP`
Are they gonna become obsolete after the merge? If they are,can you tell me where the demuxer cache will be stored(memory,internal storage or..?).Would be nice to have TMP option as i was using RAM Drive for it.

**Comment by olifre** [comment]
> If something needs to be kept it's probably best to try to single out a few things most worth saving, favoring still removing as much as possible and instead, as @qyot27 points out, trying to find ways to switch over to using ffmpeg functionality where feasible, possibly with some user-friendly short-cuts.

I agree with this. Sadly, with DVB support, there appear to be no possible alternatives that can be used - and many of the removed things are intermixed with that (e.g. TS program switching). So if this really is dropped, it means users of that have to fall back to huge blown GUI players (Kaffeine, VLC) or server solutions (tvheadend, vdr, MythTV) which *might* allow to use mpv as frontend (you could possibly run a tvheadend server and connect mpv to that somehow). 
mpv currently is the unique lightweight solution for DVB on the market, I think, so losing that - and the days I spent on that, coding and testing - would be really sad. 

Also pinging @tmm1 since he was also working on some TS / DVB related code in the past, and @bpaterni who was a heavy DVB user. If there was a way to use something existing for DVB, I'd be all for dropping it from mpv, but currently, while there are alternatives for tv / optical disc, that's not the case for DVB.

**Comment by bpaterni** [comment]
Just to echo @olifre's concern, please don't remove DVB support. We just got it working again with https://github.com/mpv-player/mpv/issues/4922 Many thanks to @olifre.I would much rather see it kept in, but if it must be removed for the sake of refactoring, please seek out and incorporate alternative external libraries, and use them if possible.

/gracious user

**Comment by jeeb** [comment]
Thanks for the PR.

It's pretty clear that this is stuff from a person's private fork (it's according to his interests). So we should start from most important bits towards the stuff of less importance/requiring further decisions (I've been meaning to look into the async command stuff/json fixups/>2GiB cache fixup etc for quite a while).

While I do recommend utilizing dvblast or something for localhost multicasts for digital TV receival, I would not remove DVB input until something like https://github.com/0p1pp1/FFmpeg/commit/924e6ce1dcd38d5a15f3bb081916f4f535db090b goes into FFmpeg.

Analog video through v4l2 seems like a mess, and we do seem to have libavdevice support for it already (libavdevice is a mess, but I think it's more towards '_how well does it work, if at all?_')

libbluray I'm not 100% sure, I should test again how well it works through lavf and it has the limitations of a protocol through not being a demuxer (I do actually use libbluray every now and then with the discs I have).

I do know from my previous testing of some other changes in the mpv libbluray parts noted that it had pretty obvious bugs regarding switching things first setting variables and then promptly resetting the input and thus going back to the default/what was set in the command line. Unfortunately I didn't get to having time to fix those issues yet.

**Comment by olifre** [comment]
> I would not remove DVB input until something like [0p1pp1/FFmpeg@924e6ce](https://github.com/0p1pp1/FFmpeg/commit/924e6ce1dcd38d5a15f3bb081916f4f535db090b) goes into FFmpeg.

That looks very promising! Also, thanks for the pointer to dvblast. I missed that, since it's not really packaged for most distros, and it seems to not have any classic channel list support which makes it significantly more uncomfortable to tune with, but it seems rather compact and powerful. 

I see a few downsides with the linked ffmpeg-implementation (but discussing this may now lead off-topic):
- No channel switching, since the player never learns about the actual channels. 
- Also, full transponder dumps don't work (which is needed for some channels which do mid-stream PID changes, I believe). 
- No automatic card selection depending on channel delivery type matched with card delivery types. 
- A lot of logic is moved from sensible automatic detection (e.g. switching the card between `DVB-S` and `DVB-S2`, which varies from channel to channel) into an enriched channel list. It really makes me wonder which program produces such channel lists, the standard tools I know right now (`w_scan` etc.) are much more limited. Do you happen to know? 

From my personal point of view, this would certainly still be something to prefer over a custom implementation as we have now, in the long run. I do especially like the simplicity of the code gained from the enriched channel list. However, the channel list format seems incompatible to the plethora of existing formats and existing tools, but it's certainly more beautiful, and I'd love to see tooling be updated :+1:.

**Comment by xantoz** [comment]
Does anyone have any opinions on the removal of big endian support?
https://github.com/mpv-player/mpv/pull/6127/commits/ef526834fb8293492877107f63bb08b6d23af90b

P.S. Regarding the mpv-native v4l2 support, it has generally worked ok for me, but I've never used the tuner only various capture cards and webcams, so going through libavdevice seems to work ok as a replacement (although I might miss the convenience of some of the options to tune things like norm and resolution) D.S

**Comment by tomty89** [comment]
While I'd love mpv to be able to play CDDA, the current support seems broken enough:
```
[tom@archlinux ~]$ cd-paranoia -Q
cdparanoia III release 10.2 libcdio 2.0.0 x86_64-pc-linux-gnu
(C) 2001 Monty <monty@xiph.org> and Xiphophorus
(C) 2004, 2005, 2008 Rocky Bernstein <rocky@gnu.org>
(C) 2014 Robert Kausch <robert.kausch@freac.org>

Report bugs to bug-libcdio@gnu.org


Table of contents (audio tracks only):
track        length               begin        copy pre ch
===========================================================
  1.    20400 [04:32.00]        0 [00:00.00]    no   no  2
  2.    19875 [04:25.00]    20400 [04:32.00]    no   no  2
  3.    21450 [04:46.00]    40275 [08:57.00]    no   no  2
  4.    21975 [04:53.00]    61725 [13:43.00]    no   no  2
  5.    15225 [03:23.00]    83700 [18:36.00]    no   no  2
  6.    21975 [04:53.00]    98925 [21:59.00]    no   no  2
  7.    17325 [03:51.00]   120900 [26:52.00]    no   no  2
  8.    18150 [04:02.00]   138225 [30:43.00]    no   no  2
  9.    29025 [06:27.00]   156375 [34:45.00]    no   no  2
 10.     8025 [01:47.00]   185400 [41:12.00]    no   no  2
TOTAL  193425 [42:59.00]    (audio only)
 
[tom@archlinux ~]$ 
```
```
[tom@archlinux ~]$ mpv cdda:// -ao jack
Playing: cdda://
[cdda] Switched to track 1
 (+) Audio --aid=1 (pcm_s16le 2ch 44100Hz)
AO: [jack] 44100Hz stereo 2ch floatp
A: 00:00:15 / 00:04:31 (5%) Cache: 256s+43MB
[cdda] Switched to track 2
A: 00:00:27 / 00:08:47 (5%) Cache: 499s+85MB
[cdda] Switched to track 3
A: 00:00:40 / 00:13:35 (5%) Cache: 774s+132MB
[cdda] Switched to track 4
A: 00:03:02 / 00:17:37 (17%) Cache: 875s+159MB
[cdda] Switched to track 5
A: 00:06:25 / 00:21:00 (30%) Cache: 875s+159MB
[cdda] Switched to track 6
A: 00:11:18 / 00:25:53 (43%) Cache: 875s+159MB
[cdda] Switched to track 7
A: 00:15:09 / 00:29:44 (50%) Cache: 875s+159MB
[cdda] Switched to track 8
A: 00:19:11 / 00:33:46 (56%) Cache: 875s+159MB
[cdda] Switched to track 9
A: 00:25:38 / 00:40:13 (63%) Cache: 875s+159MB
[cdda] Switched to track 10
A: 00:42:58 / 00:42:58 (99%) Cache:  0s


Exiting... (End of file)
[tom@archlinux ~]$ 
```
(At the end of the playback some weird sound comes out as well, it's not just a glitch or noise, more like bits of the disc is played again unexpectedly.)

The worst thing is there seems to be no way in preventing the drive to spin up with libcdio-paranoia *anyway*. It's like the library has only ripping in mind but has no proper audio CD "streaming" mode like foobar2000 in Windows does (the drive would not spin up and hence never be noisy; its LED blinks constantly in a relatively low speed).

I am not sure if anyone would still like to use the support when it is in such state.
